File: ocmsdk/src/main/java/com/gigigo/orchextra/core/data/rxRepository/rxDatasource/OcmDataStoreFactory.java
package com.gigigo.orchextra.core.data.rxRepository.rxDatasource;

import android.util.Log;
import com.gigigo.orchextra.core.data.rxCache.OcmCache;
import com.gigigo.orchextra.core.domain.entities.ocm.OxSession;
import com.gigigo.orchextra.core.domain.utils.ConnectionUtils;
import orchextra.javax.inject.Inject;
import orchextra.javax.inject.Singleton;

/**
 * Created by francisco.hernandez on 23/5/17.
 */

@Singleton public class OcmDataStoreFactory {

  private static final String TAG = OcmDataStoreFactory.class.getSimpleName();

  private final OcmCloudDataStore cloudDataStore;
  private final OcmDiskDataStore diskDataStore;
  private final ConnectionUtils connectionUtils;
  private final OxSession session;

  @Inject
  public OcmDataStoreFactory(OcmCloudDataStore cloudDataStore, OcmDiskDataStore diskDataStore,
      ConnectionUtils connectionUtils, OxSession session) {
    this.cloudDataStore = cloudDataStore;
    this.diskDataStore = diskDataStore;
    this.connectionUtils = connectionUtils;
    this.session = session;
  }

  public OcmDataStore getDataStoreForMenus(boolean force) {
    OcmDataStore ocmDataStore;

    if (!connectionUtils.hasConnection()) return getDiskDataStore();

    if (force) {
      Log.i(TAG, "CLOUD - Menus");
      ocmDataStore = getCloudDataStore();
    } else {
      OcmCache cache = diskDataStore.getOcmCache();
      if (cache.isMenuCached() && !cache.isMenuExpired()) {
        Log.i(TAG, "DISK  - Menus");
        ocmDataStore = getDiskDataStore();
      } else {
        Log.i(TAG, "CLOUD - Menus");
        ocmDataStore = getCloudDataStore();
      }
    }

    return ocmDataStore;
  }

  public OcmDataStore getDataStoreForSections(boolean force, String section) {
    OcmDataStore ocmDataStore;

    if (!connectionUtils.hasConnection()) return getDiskDataStore();

    if (force) {
      Log.i(TAG, "CLOUD - Sections");
      ocmDataStore = getCloudDataStore();
    } else {
      OcmCache cache = diskDataStore.getOcmCache();
      if (cache.isSectionCached(section) && !cache.isSectionExpired(section)) {
        Log.i(TAG, "DISK  - Sections");
        ocmDataStore = getDiskDataStore();
      } else {
        Log.i(TAG, "CLOUD - Sections");
        ocmDataStore = getCloudDataStore();
      }
    }

    return ocmDataStore;
  }

  public OcmDataStore getDataStoreForDetail(boolean force, String slug) {
    OcmDataStore ocmDataStore;

    if (!connectionUtils.hasConnection()) return getDiskDataStore();

    if (force) {
      Log.i(TAG, "CLOUD - Detail");
      ocmDataStore = getCloudDataStore();
    } else {
      OcmCache cache = diskDataStore.getOcmCache();
      if (cache.isDetailCached(slug) && !cache.isDetailExpired(slug)) {
        Log.i(TAG, "DISK  - Detail");
        ocmDataStore = getDiskDataStore();
      } else {
        Log.i(TAG, "CLOUD - Detail");
        ocmDataStore = getCloudDataStore();
      }
    }

    return ocmDataStore;
  }

  public OcmDataStore getDataStoreForVideo(boolean force, String videoId, boolean isWifiConnection,
      boolean isFastConnection) {

    OcmDataStore ocmDataStore;

    if (!connectionUtils.hasConnection()) return getDiskDataStore();

    if (force) {
      Log.i(TAG, "CLOUD - Video");
      ocmDataStore = getCloudDataStore();
    } else {
      OcmCache cache = diskDataStore.getOcmCache();
      if (cache.isVideoCached(videoId) && !cache.isVideoExpired(videoId)) {
        Log.i(TAG, "DISK  - Video");
        ocmDataStore = getDiskDataStore();
      } else {
        Log.i(TAG, "CLOUD - Video");
        ocmDataStore = getCloudDataStore();
      }
    }

    return ocmDataStore;
  }

  public OcmDataStore getCloudDataStore() {
    return cloudDataStore;
  }

  public OcmDataStore getDiskDataStore() {
    return diskDataStore;
  }

  public OcmDataStore getDataStoreForVersion() {
    if (session.getToken() == null) {
      return getDiskDataStore();
    }

    OcmCache ocmCache = diskDataStore.getOcmCache();

    if (ocmCache.isVersionCached() && !ocmCache.isVersionExpired()) {
      Log.i(TAG, "DISK  - Version");
      return getDiskDataStore();
    } else {
      Log.i(TAG, "CLOUD - Version");
      return getCloudDataStore();
    }
  }
}

File: ocmsdk/src/main/java/com/gigigo/orchextra/core/data/rxRepository/rxDatasource/OcmDataStoreFactory.kt
package com.gigigo.orchextra.core.data.rxRepository.rxDatasource

import android.util.Log
import com.gigigo.orchextra.core.domain.entities.ocm.OxSession
import com.gigigo.orchextra.core.domain.utils.ConnectionUtils
import com.gigigo.orchextra.core.utils.CoroutineExecutor
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.CoroutineContext


private val uiContext: CoroutineContext = Dispatchers.Main
private val bgContext: CoroutineContext = CoroutineExecutor.BACKGROUND

@Singleton
class OcmDataStoreFactory
@Inject constructor(
    val cloudDataStore: OcmCloudDataStore,
    val diskDataStore: OcmDiskDataStore,
    private val connectionUtils: ConnectionUtils,
    private val session: OxSession
) {

    private val TAG = OcmDataStoreFactory::class.java.simpleName

    fun getDataStoreForVersion(): OcmDataStore {
        if (session.token == null) {
            return diskDataStore
        }

        lateinit var dataStore: OcmDataStore

        runBlocking {
            println("*****GETVERSION RUNBLOCKING THREAD " + Thread.currentThread().name)
            dataStore = dataStoreForVersion().await()
        }

        return dataStore
    }

    private fun dataStoreForVersion() =
        GlobalScope.async(bgContext) {

            println("*****GETVERSION ASYNC THREAD " + Thread.currentThread().name)
            var dataStore: OcmDataStore

            if (false) {
                Log.i(TAG, "DISK - Version")
                dataStore = diskDataStore
            } else {
                Log.i(TAG, "CLOUD - Version")
                dataStore = cloudDataStore
            }

            dataStore
        }

    fun getDataStoreForMenus(force: Boolean): OcmDataStore {
        //if (!connectionUtils.hasConnection()) return diskDataStore

        lateinit var dataStore: OcmDataStore

        runBlocking {
            dataStore = dataStoreForMenus(force).await()
        }

        return dataStore
    }

    private fun dataStoreForMenus(force: Boolean) =
        GlobalScope.async(bgContext) {
            var dataStore: OcmDataStore

            dataStore = if (force) {
                Log.i(TAG, "CLOUD - Menus")
                cloudDataStore
            } else {
                val cache = diskDataStore.ocmCache
                if (cache.hasMenusCached()) {
                    Log.i(TAG, "DISK  - Menus")
                    diskDataStore
                } else {
                    Log.i(TAG, "CLOUD - Menus")
                    cloudDataStore
                }
            }

            dataStore
        }

    fun getDataStoreForSection(force: Boolean, section: String): OcmDataStore {
        if (!connectionUtils.hasConnection()) return diskDataStore

        lateinit var dataStore: OcmDataStore

        runBlocking {
            dataStore = dataStoreForSection(force, section).await()
        }

        return dataStore
    }

    private fun dataStoreForSection(force: Boolean, section: String) =
        GlobalScope.async(bgContext) {
            var dataStore: OcmDataStore

            dataStore = if (force) {
                Log.i(TAG, "CLOUD - Sections")
                cloudDataStore
            } else {
                val cache = diskDataStore.ocmCache
                if (cache.isSectionCached(section) && !cache.isSectionExpired(section)) {
                    Log.i(TAG, "DISK  - Sections")
                    diskDataStore
                } else {
                    Log.i(TAG, "CLOUD - Sections")
                    cloudDataStore
                }
            }

            dataStore
        }

    fun getDataStoreForDetail(force: Boolean, slug: String): OcmDataStore {
        if (!connectionUtils.hasConnection()) return diskDataStore

        lateinit var dataStore: OcmDataStore

        runBlocking {
            dataStore = dataStoreForDetail(force, slug).await()
        }

        return dataStore
    }

    private fun dataStoreForDetail(force: Boolean, slug: String) =
        GlobalScope.async(bgContext) {
            var dataStore: OcmDataStore

            dataStore = if (force) {
                Log.i(TAG, "CLOUD - Detail")
                cloudDataStore
            } else {
                val cache = diskDataStore.ocmCache
                if (cache.isDetailCached(slug) && !cache.isDetailExpired(slug)) {
                    Log.i(TAG, "DISK  - Detail")
                    diskDataStore
                } else {
                    Log.i(TAG, "CLOUD - Detail")
                    cloudDataStore
                }
            }

            dataStore
        }

    fun getDataStoreForVideo(force: Boolean, videoId: String): OcmDataStore {
        if (!connectionUtils.hasConnection()) return diskDataStore

        lateinit var dataStore: OcmDataStore

        runBlocking {
            dataStore = dataStoreForVideo(force, videoId).await()
        }

        return dataStore
    }

    private fun dataStoreForVideo(force: Boolean, videoId: String) =
        GlobalScope.async(bgContext) {
            var dataStore: OcmDataStore

            dataStore = if (force) {
                Log.i(TAG, "CLOUD - Video")
                cloudDataStore
            } else {
                val cache = diskDataStore.ocmCache
                if (cache.isVideoCached(videoId) && !cache.isVideoExpired(videoId)) {
                    Log.i(TAG, "DISK  - Video")
                    diskDataStore
                } else {
                    Log.i(TAG, "CLOUD - Video")
                    cloudDataStore
                }
            }

            dataStore
        }
}